src/main/java/gtPlusPlus/core/util/recipe/GT_RecipeUtils.java

```diff
@@ -0,0 +1,98 @@
+                } else {
+                    savedCircuit = itemStack;
+                }
+            }
+            GT_Recipe newRecipe = new GT_Recipe(
+                    false,
+                    ItemInputsWithoutProgrammableCircuit.toArray(new ItemStack[0]),
+                    recipeInput.mOutputs,
+                    recipeInput.mSpecialItems,
+                    recipeInput.mChances,
+                    recipeInput.mFluidInputs,
+                    recipeInput.mFluidOutputs,
+                    recipeInput.mDuration,
+                    recipeInput.mEUt,
+                    recipeInput.mSpecialValue);
+            if (!recipes.contains(newRecipe)) {
+                // if the recipes customHashSet does not contain the new recipe then add it
+                recipes.add(newRecipe);
+            } else {
+                // add the removed recipe to the removal array with its original item inputs.
+                GT_Recipe recipeCopy = newRecipe.copy();
+                recipeCopy.mInputs = recipeInput.mInputs;
+                removedRecipes.add(recipeCopy);
+            }
+            if (savedCircuit != null) {
+                // if the current recipe has a circuit and the recipe (without circuits) is already in the
+                // circuit map then check make sure the circuit map saves the recipe with the smallest circuit
+                // damage value. This is to prevent a case where recipe load order would affect which duplicate
+                // recipes with multiple circuit values gets removed.
+                if (circuitMap.containsKey(newRecipe)) {
+                    if (circuitMap.get(newRecipe).getItemDamage() > savedCircuit.getItemDamage()) {
+                        circuitMap.put(newRecipe, savedCircuit);
+                    }
+                } else {
+                    // If the circuit map does not have the recipe in it yet then add it
+                    circuitMap.put(newRecipe, savedCircuit);
+                }
+            }
+        }
+        // iterate over all recipes without duplicates and add them to the output. If the recipe had a programmable
+        // circuit in it then add it back with its damage value coming from the circuit map.
+        for (GT_Recipe filteredRecipe : recipes.toArray(new GT_Recipe[0])) {
+            // check to see if the recipe is in the circuit map
+            if (circuitMap.contains(filteredRecipe)) {
+                ArrayList<ItemStack> revertedInputItemStackArray = new ArrayList<>(
+                        Arrays.asList(filteredRecipe.mInputs));
+                // add the circuit back
+                revertedInputItemStackArray.add(circuitMap.get(filteredRecipe));
+                // update the item input array with the new input from
+                // ItemInputsWithoutProgrammableCircuit + circuit map circuit
+                filteredRecipe.mInputs = revertedInputItemStackArray.toArray(new ItemStack[0]);
```

Comment:
`ArrayUtils.add` is probably suitable for this task, rather than converting between array <-> list.